File: deepspeed/inference/v2/kernels/ragged_ops/embed/embed.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

from typing import Optional

import torch

from ... import DSKernelBase
from deepspeed.ops.op_builder import RaggedOpsBuilder
from ....inference_utils import elem_size
from ....ragged import RaggedBatchWrapper


class RaggedEmbeddingKernel(DSKernelBase):
    """
    Ragged-aware CUDA kernel implementation for an embedding lookup. This will only lookup
    the necessary tokens for a padded batch (i.e. if we are CGed and running with a slightly
    larger batch size than the actual tokens).
    """

    supported_dtypes = [torch.float16, torch.bfloat16, torch.float32]
    supported_token_dtypes = [torch.int32, torch.int64]

    def __init__(self, embed_dtype: torch.dtype, token_dtype: torch.dtype, embed_dim: int) -> None:
        """
        Args:
            fp_dtype (torch.dtype): Data type of the embedding table and output dtype.
                Supported values are torch.float16, torch.bfloat16, and torch.float32.
            token_dtype (torch.dtype): Data type of the token ids. Supported values are
                torch.int32 and torch.int64.
            embed_dim (int): Embedding dimension. Must be aligned to 16 bytes.
        """
        if embed_dtype not in RaggedEmbeddingKernel.supported_dtypes:
            raise ValueError("Unsupported embedding data type: {}, supported_dtypes are {}".format(
                embed_dtype, RaggedEmbeddingKernel.supported_dtypes))

        if token_dtype not in RaggedEmbeddingKernel.supported_token_dtypes:
            raise ValueError("Unsupported token data type: {}, supported_dtypes are {}".format(
                token_dtype, RaggedEmbeddingKernel.supported_token_dtypes))

        if elem_size(embed_dtype) * embed_dim % 16 != 0:
            raise ValueError("Embedding dimension must be aligned to 16 bytes, got {}".format(embed_dim))

        inf_module = RaggedOpsBuilder().load()
        self.kernel = inf_module.ragged_embed

    def __call__(self,
                 embedded_tokens: torch.Tensor,
                 ragged_wrapper: RaggedBatchWrapper,
                 embedding_weight: torch.Tensor,
                 position_embed_weight: Optional[torch.Tensor] = None,
                 position_embed_offset: int = 0) -> torch.Tensor:
        """
        Ragged aware embedding lookup.

        Args:
            embedded_tokens (torch.Tensor): Output tensor of shape [num_tokens, embed_dim]
            ragged_wrapper (RaggedBatchWrapper): Wrapper for the ragged batch.
            embedding_weight (torch.Tensor): Embedding table of shape [vocab_size, embed_dim]
        """
        self.kernel(embedded_tokens, ragged_wrapper.input_ids(),
                    embedding_weight, position_embed_weight, position_embed_offset,
                    ragged_wrapper.batch_metadata_buffer(), ragged_wrapper.inflight_seq_descriptors(),
                    ragged_wrapper.tokens_to_seq(), ragged_wrapper.kv_ptrs())
        return embedded_tokens

File: deepspeed/inference/v2/kernels/ragged_ops/embed/embed.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

#include "embed.h"
#include "ragged_kernel_helpers.h"

#ifdef BF16_AVAILABLE
#define DISPATCH_FOR_FLOAT(DTYPE, ...)                       \
    [&] {                                                    \
        if (DTYPE == torch::kFloat32) {                      \
            using float_t = float;                           \
            return __VA_ARGS__();                            \
        } else if (DTYPE == torch::kFloat16) {               \
            using float_t = __half;                          \
            return __VA_ARGS__();                            \
        } else if (DTYPE == torch::kBFloat16) {              \
            using float_t = __nv_bfloat16;                   \
            return __VA_ARGS__();                            \
        } else {                                             \
            TORCH_CHECK(false, "Unsupported dispatch type"); \
        }                                                    \
    }()
#else
#define DISPATCH_FOR_FLOAT(DTYPE, ...)                       \
    [&] {                                                    \
        if (DTYPE == torch::kFloat32) {                      \
            using float_t = float;                           \
            return __VA_ARGS__();                            \
        } else if (DTYPE == torch::kFloat16) {               \
            using float_t = __half;                          \
            return __VA_ARGS__();                            \
        } else {                                             \
            TORCH_CHECK(false, "Unsupported dispatch type"); \
        }                                                    \
    }()
#endif

#define DISPATCH_FOR_INT(DTYPE, ...)                         \
    [&] {                                                    \
        if (DTYPE == torch::kInt32) {                        \
            using int_t = int32_t;                           \
            return __VA_ARGS__();                            \
        } else if (DTYPE == torch::kInt64) {                 \
            using int_t = int64_t;                           \
            return __VA_ARGS__();                            \
        } else {                                             \
            TORCH_CHECK(false, "Unsupported dispatch type"); \
        }                                                    \
    }()

/*
Embeddings kernel aware of ragged batch structure.
*/
void ragged_embed(torch::Tensor& embedded_tokens,
                  torch::Tensor& input_ids,
                  torch::Tensor& embedding_weight,
                  c10::optional<torch::Tensor>& position_embedding_weight,
                  int32_t pos_embed_offset,
                  torch::Tensor& batch_metadata,
                  torch::Tensor& seq_metadata,
                  torch::Tensor& tokens_to_seq,
                  torch::Tensor& kv_ptrs)
{
    // We don't care about KV cache here, so just hardcoding 0s for block_size/num_blocks
    BatchWrapperCPP batch_wrapper =
        make_cpp_batch_wrapper(batch_metadata, seq_metadata, tokens_to_seq, kv_ptrs, 0, 0);

    const int32_t n_tokens = input_ids.numel();
    const int32_t embed_dim = embedding_weight.size(1);
    const int32_t vocab_size = embedding_weight.size(0);

    DISPATCH_FOR_INT(input_ids.scalar_type(), [&] {
        DISPATCH_FOR_FLOAT(embedding_weight.scalar_type(), [&] {
            float_t* pos_embed_ptr = nullptr;
            int32_t max_position_embed_idx = 0;
            if (position_embedding_weight.has_value()) {
                TORCH_CHECK(
                    position_embedding_weight.value().options().dtype() ==
                        embedding_weight.options().dtype(),
                    "position_embedding_weight and embedding_weight must have the same dtype");
                pos_embed_ptr =
                    reinterpret_cast<float_t*>(position_embedding_weight.value().data_ptr());
                max_position_embed_idx = position_embedding_weight.value().size(0) - 1;
            }

            launch_ragged_embed_kernel((float_t*)embedded_tokens.data_ptr(),
                                       (const int_t*)input_ids.data_ptr(),
                                       (const float_t*)embedding_weight.data_ptr(),
                                       pos_embed_ptr,
                                       batch_wrapper,
                                       n_tokens,
                                       embed_dim,
                                       vocab_size,
                                       max_position_embed_idx,
                                       pos_embed_offset,
                                       at::cuda::getCurrentCUDAStream());
        });
    });
}
